[Bug]: Broken gRPC build with VS2019

### Describe the issue

When building gRPC with bazel RBE, I'm getting the `absl/types/compare.h(49): error C2338: Only literal "0" is allowed.` error

The issue started happening after upgrading to protobuf 22.x, but the error seems to be coming from absl codebase.
Perhaps this is some incompatibility with C++14 and MSVC from VS2019?

```
ERROR: T:/altsrc/github/grpc/test/core/xds/BUILD:232:13: Compiling test/core/xds/xds_listener_resource_type_test.cc failed: (Exit 2): cl.exe failed: error executing command
  cd /d T:/_bazel_output/pjdymepj/execroot/com_github_grpc_grpc
  SET BAZEL_DO_NOT_DETECT_CPP_TOOLCHAIN=1
    SET GRPC_BAZEL_RUNTIME=1
    SET INCLUDE=C:\VS\VC\Tools\MSVC\14.29.30133\include;C:\Program Files (x86)\Windows Kits\10\include\10.0.19041.0\ucrt;C:\Program Files (x86)\Windows Kits\10\include\10.0.19041.0\shared;C:\Program Files (x86)\Windows Kits\10\include\10.0.19041.0\um;C:\Program Files (x86)\Windows Kits\10\include\10.0.19041.0\winrt;C:\Program Files (x86)\Windows Kits\10\include\10.0.19041.0\cppwinrt
    SET PATH=C:\VS\VC\Tools\MSVC\14.29.30133\bin\HostX64\x64;C:\VS\Common7\IDE\VC\VCPackages;C:\VS\Common7\IDE\CommonExtensions\Microsoft\TestWindow;C:\VS\Common7\IDE\CommonExtensions\Microsoft\TeamFoundation\Team Explorer;C:\VS\MSBuild\Current\bin\Roslyn;C:\VS\Common7\Tools\devinit;C:\Program Files (x86)\Windows Kits\10\bin\10.0.19041.0\x64;C:\Program Files (x86)\Windows Kits\10\bin\x64;C:\VS\\MSBuild\Current\Bin;C:\Windows\Microsoft.NET\Framework64\v4.0.30319;C:\VS\Common7\IDE\;C:\VS\Common7\Tools\;;C:\Windows\system32
    SET PWD=/proc/self/cwd
    SET RUNFILES_MANIFEST_ONLY=1
    SET TEMP=C:\Users\ContainerAdministrator\AppData\Local\Temp
    SET TMP=C:\Users\ContainerAdministrator\AppData\Local\Temp
  C:\VS\VC\Tools\MSVC\14.29.30133\bin\HostX64\x64\cl.exe /nologo /DCOMPILER_MSVC /DNOMINMAX /D_WIN32_WINNT=0x0601 /D_CRT_SECURE_NO_DEPRECATE /D_CRT_SECURE_NO_WARNINGS /bigobj /Zm500 /EHsc /wd4351 /wd4291 /wd4250 /wd4996 /I. /Ibazel-out/x64_windows-opt/bin /Iexternal/com_google_absl /Ibazel-out/x64_windows-opt/bin/external/com_google_absl /Iexternal/com_google_googleapis /Ibazel-out/x64_windows-opt/bin/external/com_google_googleapis /Iexternal/upb /Ibazel-out/x64_windows-opt/bin/external/upb /Iexternal/utf8_range /Ibazel-out/x64_windows-opt/bin/external/utf8_range /Iexternal/com_google_protobuf /Ibazel-out/x64_windows-opt/bin/external/com_google_protobuf /Iexternal/zlib /Ibazel-out/x64_windows-opt/bin/external/zlib /Iexternal/boringssl /Ibazel-out/x64_windows-opt/bin/external/boringssl /Iexternal/com_github_cncf_udpa /Ibazel-out/x64_windows-opt/bin/external/com_github_cncf_udpa /Iexternal/com_envoyproxy_protoc_gen_validate /Ibazel-out/x64_windows-opt/bin/external/com_envoyproxy_protoc_gen_validate /Iexternal/com_github_cares_cares /Ibazel-out/x64_windows-opt/bin/external/com_github_cares_cares /Iexternal/com_googlesource_code_re2 /Ibazel-out/x64_windows-opt/bin/external/com_googlesource_code_re2 /Iexternal/envoy_api /Ibazel-out/x64_windows-opt/bin/external/envoy_api /Iexternal/opencensus_proto /Ibazel-out/x64_windows-opt/bin/external/opencensus_proto /Iexternal/com_google_googletest /Ibazel-out/x64_windows-opt/bin/external/com_google_googletest /Iexternal/bazel_tools /Ibazel-out/x64_windows-opt/bin/external/bazel_tools /Ibazel-out/x64_windows-opt/bin/external/com_google_googleapis /Ibazel-out/x64_windows-opt/bin/external/com_google_protobuf/src /Ibazel-out/x64_windows-opt/bin /Ibazel-out/x64_windows-opt/bin/external/com_github_cncf_udpa /Ibazel-out/x64_windows-opt/bin/external/com_envoyproxy_protoc_gen_validate /Ibazel-out/x64_windows-opt/bin/external/envoy_api /Ibazel-out/x64_windows-opt/bin/external/opencensus_proto /Ibazel-out/x64_windows-opt/bin/external/com_google_protobuf/_virtual_includes/protobuf /Ibazel-out/x64_windows-opt/bin/external/com_google_protobuf/src/google/protobuf/_virtual_includes/protobuf /Ibazel-out/x64_windows-opt/bin/external/com_google_protobuf/src/google/protobuf/_virtual_includes/protobuf_nowkt /Ibazel-out/x64_windows-opt/bin/external/com_google_protobuf/src/google/protobuf/_virtual_includes/protobuf_lite /Ibazel-out/x64_windows-opt/bin/external/com_google_protobuf/src/google/protobuf/_virtual_includes/arena /Ibazel-out/x64_windows-opt/bin/external/com_google_protobuf/src/google/protobuf/_virtual_includes/arena_align /Ibazel-out/x64_windows-opt/bin/external/com_google_protobuf/src/google/protobuf/stubs/_virtual_includes/lite /Ibazel-out/x64_windows-opt/bin/external/com_google_protobuf/src/google/protobuf/_virtual_includes/port_def /Ibazel-out/x64_windows-opt/bin/external/com_google_protobuf/src/google/protobuf/_virtual_includes/arena_allocation_policy /Ibazel-out/x64_windows-opt/bin/external/com_google_protobuf/src/google/protobuf/_virtual_includes/arena_config /Ibazel-out/x64_windows-opt/bin/external/com_google_protobuf/src/google/protobuf/_virtual_includes/arena_cleanup /Ibazel-out/x64_windows-opt/bin/external/com_google_protobuf/src/google/protobuf/_virtual_includes/string_block /Ibazel-out/x64_windows-opt/bin/external/com_google_protobuf/src/google/protobuf/io/_virtual_includes/io /Ibazel-out/x64_windows-opt/bin/external/com_google_protobuf/src/google/protobuf/io/_virtual_includes/io_win32 /Ibazel-out/x64_windows-opt/bin/external/com_google_protobuf/src/google/protobuf/io/_virtual_includes/gzip_stream /Ibazel-out/x64_windows-opt/bin/external/com_google_protobuf/src/google/protobuf/stubs/_virtual_includes/stubs /Ibazel-out/x64_windows-opt/bin/external/com_google_protobuf/src/google/protobuf/io/_virtual_includes/printer /Ibazel-out/x64_windows-opt/bin/external/com_google_protobuf/src/google/protobuf/io/_virtual_includes/zero_copy_sink /Ibazel-out/x64_windows-opt/bin/external/com_google_protobuf/src/google/protobuf/io/_virtual_includes/tokenizer /Ibazel-out/x64_windows-opt/bin/external/com_google_protobuf/src/google/protobuf/compiler/_virtual_includes/importer /Ibazel-out/x64_windows-opt/bin/external/com_google_protobuf/src/google/protobuf/util/_virtual_includes/delimited_message_util /Ibazel-out/x64_windows-opt/bin/external/com_google_protobuf/src/google/protobuf/util/_virtual_includes/differencer /Ibazel-out/x64_windows-opt/bin/external/com_google_protobuf/src/google/protobuf/util/_virtual_includes/field_mask_util /Ibazel-out/x64_windows-opt/bin/external/com_google_protobuf/src/google/protobuf/util/_virtual_includes/json_util /Ibazel-out/x64_windows-opt/bin/external/com_google_protobuf/src/google/protobuf/json/_virtual_includes/json /Ibazel-out/x64_windows-opt/bin/external/com_google_protobuf/src/google/protobuf/json/_virtual_includes/parser /Ibazel-out/x64_windows-opt/bin/external/com_google_protobuf/src/google/protobuf/json/_virtual_includes/descriptor_traits /Ibazel-out/x64_windows-opt/bin/external/com_google_protobuf/src/google/protobuf/json/_virtual_includes/lexer /Ibazel-out/x64_windows-opt/bin/external/com_google_protobuf/src/google/protobuf/json/_virtual_includes/message_path /Ibazel-out/x64_windows-opt/bin/external/com_google_protobuf/src/google/protobuf/json/_virtual_includes/zero_copy_buffered_stream /Ibazel-out/x64_windows-opt/bin/external/com_google_protobuf/src/google/protobuf/json/_virtual_includes/untyped_message /Ibazel-out/x64_windows-opt/bin/external/com_google_protobuf/src/google/protobuf/util/_virtual_includes/type_resolver_util /Ibazel-out/x64_windows-opt/bin/external/com_google_protobuf/src/google/protobuf/json/_virtual_includes/unparser /Ibazel-out/x64_windows-opt/bin/external/com_google_protobuf/src/google/protobuf/json/_virtual_includes/writer /Ibazel-out/x64_windows-opt/bin/external/com_google_protobuf/src/google/protobuf/util/_virtual_includes/time_util /Ibazel-out/x64_windows-opt/bin/external/com_google_protobuf/src/google/protobuf/_virtual_includes/any_proto /Ibazel-out/x64_windows-opt/bin/external/com_google_protobuf/src/google/protobuf/_virtual_includes/api_proto /Ibazel-out/x64_windows-opt/bin/external/com_google_protobuf/src/google/protobuf/_virtual_includes/source_context_proto /Ibazel-out/x64_windows-opt/bin/external/com_google_protobuf/src/google/protobuf/_virtual_includes/type_proto /Ibazel-out/x64_windows-opt/bin/external/com_google_protobuf/src/google/protobuf/compiler/_virtual_includes/plugin_proto /Ibazel-out/x64_windows-opt/bin/external/com_google_protobuf/src/google/protobuf/_virtual_includes/descriptor_proto /Ibazel-out/x64_windows-opt/bin/external/com_google_protobuf/src/google/protobuf/_virtual_includes/duration_proto /Ibazel-out/x64_windows-opt/bin/external/com_google_protobuf/src/google/protobuf/_virtual_includes/empty_proto /Ibazel-out/x64_windows-opt/bin/external/com_google_protobuf/src/google/protobuf/_virtual_includes/field_mask_proto /Ibazel-out/x64_windows-opt/bin/external/com_google_protobuf/src/google/protobuf/_virtual_includes/struct_proto /Ibazel-out/x64_windows-opt/bin/external/com_google_protobuf/src/google/protobuf/_virtual_includes/timestamp_proto /Ibazel-out/x64_windows-opt/bin/external/com_google_protobuf/src/google/protobuf/_virtual_includes/wrappers_proto /Ibazel-out/x64_windows-opt/bin/external/com_google_protobuf/src/google/protobuf/compiler/_virtual_includes/code_generator /Ibazel-out/x64_windows-opt/bin/external/com_google_protobuf/src/google/protobuf/compiler/_virtual_includes/command_line_interface /Ibazel-out/x64_windows-opt/bin/external/com_google_protobuf/src/google/protobuf/compiler/cpp/_virtual_includes/cpp /Ibazel-out/x64_windows-opt/bin/external/com_google_protobuf/src/google/protobuf/compiler/cpp/_virtual_includes/names /Ibazel-out/x64_windows-opt/bin/external/com_google_protobuf/src/google/protobuf/compiler/cpp/_virtual_includes/names_internal /Ibazel-out/x64_windows-opt/bin/external/com_google_protobuf/src/google/protobuf/compiler/_virtual_includes/retention /Ibazel-out/x64_windows-opt/bin/external/com_google_protobuf/src/google/protobuf/compiler/csharp/_virtual_includes/csharp /Ibazel-out/x64_windows-opt/bin/external/com_google_protobuf/src/google/protobuf/compiler/csharp/_virtual_includes/names /Ibazel-out/x64_windows-opt/bin/external/com_google_protobuf/src/google/protobuf/compiler/java/_virtual_includes/java /Ibazel-out/x64_windows-opt/bin/external/com_google_protobuf/src/google/protobuf/compiler/java/_virtual_includes/names /Ibazel-out/x64_windows-opt/bin/external/com_google_protobuf/src/google/protobuf/compiler/java/_virtual_includes/names_internal /Ibazel-out/x64_windows-opt/bin/external/com_google_protobuf/src/google/protobuf/compiler/objectivec/_virtual_includes/objectivec /Ibazel-out/x64_windows-opt/bin/external/com_google_protobuf/src/google/protobuf/compiler/objectivec/_virtual_includes/line_consumer /Ibazel-out/x64_windows-opt/bin/external/com_google_protobuf/src/google/protobuf/compiler/objectivec/_virtual_includes/names /Ibazel-out/x64_windows-opt/bin/external/com_google_protobuf/src/google/protobuf/compiler/objectivec/_virtual_includes/names_internal /Ibazel-out/x64_windows-opt/bin/external/com_google_protobuf/src/google/protobuf/compiler/php/_virtual_includes/php /Ibazel-out/x64_windows-opt/bin/external/com_google_protobuf/src/google/protobuf/compiler/php/_virtual_includes/names /Ibazel-out/x64_windows-opt/bin/external/com_google_protobuf/src/google/protobuf/compiler/python/_virtual_includes/python /Ibazel-out/x64_windows-opt/bin/external/com_google_protobuf/src/google/protobuf/compiler/ruby/_virtual_includes/ruby /Iinclude /Ibazel-out/x64_windows-opt/bin/include /Isrc/core/ext/upb-generated /Ibazel-out/x64_windows-opt/bin/src/core/ext/upb-generated /Isrc/core/ext/upbdefs-generated /Ibazel-out/x64_windows-opt/bin/src/core/ext/upbdefs-generated /Isrc/core/include /Ibazel-out/x64_windows-opt/bin/src/core/include /Isrc/core/src/core/ext/upb-generated /Ibazel-out/x64_windows-opt/bin/src/core/src/core/ext/upb-generated /Isrc/core/src/core/ext/upbdefs-generated /Ibazel-out/x64_windows-opt/bin/src/core/src/core/ext/upbdefs-generated /Iexternal/zlib /Ibazel-out/x64_windows-opt/bin/external/zlib /Iexternal/boringssl/src/include /Ibazel-out/x64_windows-opt/bin/external/boringssl/src/include /Ithird_party/xxhash /Ibazel-out/x64_windows-opt/bin/third_party/xxhash /Ithird_party/address_sorting/include /Ibazel-out/x64_windows-opt/bin/third_party/address_sorting/include /Iexternal/com_github_cares_cares/include /Ibazel-out/x64_windows-opt/bin/external/com_github_cares_cares/include /Iexternal/com_github_cares_cares /Ibazel-out/x64_windows-opt/bin/external/com_github_cares_cares /Iexternal/com_google_protobuf/src/google/protobuf/wkt /Ibazel-out/x64_windows-opt/bin/external/com_google_protobuf/src/google/protobuf/wkt /Iexternal/com_google_googletest/googlemock /Ibazel-out/x64_windows-opt/bin/external/com_google_googletest/googlemock /Iexternal/com_google_googletest/googlemock/include /Ibazel-out/x64_windows-opt/bin/external/com_google_googletest/googlemock/include /Iexternal/com_google_googletest/googletest /Ibazel-out/x64_windows-opt/bin/external/com_google_googletest/googletest /Iexternal/com_google_googletest/googletest/include /Ibazel-out/x64_windows-opt/bin/external/com_google_googletest/googletest/include /Itest/core/util/include /Ibazel-out/x64_windows-opt/bin/test/core/util/include /Itest/core/util/src/core/ext/upb-generated /Ibazel-out/x64_windows-opt/bin/test/core/util/src/core/ext/upb-generated /Itest/core/util/src/core/ext/upbdefs-generated /Ibazel-out/x64_windows-opt/bin/test/core/util/src/core/ext/upbdefs-generated /Itest/core/event_engine/include /Ibazel-out/x64_windows-opt/bin/test/core/event_engine/include /Itest/core/event_engine/src/core/ext/upb-generated /Ibazel-out/x64_windows-opt/bin/test/core/event_engine/src/core/ext/upb-generated /Itest/core/event_engine/src/core/ext/upbdefs-generated /Ibazel-out/x64_windows-opt/bin/test/core/event_engine/src/core/ext/upbdefs-generated /Itest/cpp/util/include /Ibazel-out/x64_windows-opt/bin/test/cpp/util/include /Itest/cpp/util/src/core/ext/upb-generated /Ibazel-out/x64_windows-opt/bin/test/cpp/util/src/core/ext/upb-generated /Itest/cpp/util/src/core/ext/upbdefs-generated /Ibazel-out/x64_windows-opt/bin/test/cpp/util/src/core/ext/upbdefs-generated /DGRPC_PORT_ISOLATED_RUNTIME=1 /DCARES_STATICLIB /showIncludes /MD /O2 /Oy- /DNDEBUG /wd4117 -D__DATE__="redacted" -D__TIMESTAMP__="redacted" -D__TIME__="redacted" /Gy /Gw -DGRPC_BAZEL_BUILD /std:c++14 /Fobazel-out/x64_windows-opt/bin/test/core/xds/_objs/xds_listener_resource_type_test/xds_listener_resource_type_test.obj /c test/core/xds/xds_listener_resource_type_test.cc
# Configuration: 01161858ca5431053053b80434f2c1a48dbf382dff4b7d989aada28be8a11818
# Execution platform: //third_party/toolchains:rbe_windows_default_platform
external/com_google_absl\absl/types/compare.h(49): error C2338: Only literal `0` is allowed.
external/com_google_absl\absl/types/optional.h(704): note: see reference to class template instantiation 'absl::lts_20230125::compare_internal::Fail<T>' being compiled
        with
        [
            T=uint32_t
        ]
```

https://source.cloud.google.com/results/invocations/07335c45-6d16-44d8-bf4b-3b92d6472328/log

### Steps to reproduce the problem

Build gRPC with bazel 

### What version of Abseil are you using?

abseil-cpp-20230125.0

### What operating system and version are you using?

Windows 2019 Server

### What compiler and version are you using?

MSVC from VS2019

### What build system are you using?

bazel

### Additional context

_No response_

In addition to the original report, I'm seeing the same problem when building with VS2017 with cmake.

```
[6083/6128] Building CXX object CMakeFiles\xds_listener_resource_type_test.dir\test\core\xds\xds_listener_resource_type_test.cc.obj

FAILED: CMakeFiles/xds_listener_resource_type_test.dir/test/core/xds/xds_listener_resource_type_test.cc.obj

C:\ccache\ccache.exe C:\PROGRA~2\MICROS~1\2017\COMMUN~1\VC\Tools\MSVC\1416~1.270\bin\Hostx64\x86\cl.exe  /nologo /TP -DCARES_STATICLIB -D_CRT_SECURE_NO_WARNINGS -D_SCL_SECURE_NO_WARNINGS -D_WIN32_WINNT=0x600 -D_WINSOCK_DEPRECATED_NO_WARNINGS -IT:\altsrc\github\grpc\workspace_c++_windows_dbg_native_default_default_buildonly\third_party\re2 -IT:\altsrc\github\grpc\workspace_c++_windows_dbg_native_default_default_buildonly\third_party\zlib -IT:\altsrc\github\grpc\workspace_c++_windows_dbg_native_default_default_buildonly -IT:\altsrc\github\grpc\workspace_c++_windows_dbg_native_default_default_buildonly\include -IT:\altsrc\github\grpc\workspace_c++_windows_dbg_native_default_default_buildonly\third_party\address_sorting\include -IT:\altsrc\github\grpc\workspace_c++_windows_dbg_native_default_default_buildonly\cmake\build\third_party\re2 -IT:\altsrc\github\grpc\workspace_c++_windows_dbg_native_default_default_buildonly\third_party\boringssl-with-bazel\src\include -IT:\altsrc\github\grpc\workspace_c++_windows_dbg_native_default_default_buildonly\src\core\ext\upb-generated -IT:\altsrc\github\grpc\workspace_c++_windows_dbg_native_default_default_buildonly\src\core\ext\upbdefs-generated -IT:\altsrc\github\grpc\workspace_c++_windows_dbg_native_default_default_buildonly\third_party\upb -IT:\altsrc\github\grpc\workspace_c++_windows_dbg_native_default_default_buildonly\third_party\utf8_range -IT:\altsrc\github\grpc\workspace_c++_windows_dbg_native_default_default_buildonly\third_party\xxhash -IT:\altsrc\github\grpc\workspace_c++_windows_dbg_native_default_default_buildonly\cmake\build\third_party\zlib -IT:\altsrc\github\grpc\workspace_c++_windows_dbg_native_default_default_buildonly\third_party\googletest\googletest\include -IT:\altsrc\github\grpc\workspace_c++_windows_dbg_native_default_default_buildonly\third_party\googletest\googletest -IT:\altsrc\github\grpc\workspace_c++_windows_dbg_native_default_default_buildonly\third_party\googletest\googlemock\include -IT:\altsrc\github\grpc\workspace_c++_windows_dbg_native_default_default_buildonly\third_party\googletest\googlemock -IT:\altsrc\github\grpc\workspace_c++_windows_dbg_native_default_default_buildonly\cmake\build\gens -IT:\altsrc\github\grpc\workspace_c++_windows_dbg_native_default_default_buildonly\third_party\protobuf\src -IT:\altsrc\github\grpc\workspace_c++_windows_dbg_native_default_default_buildonly\third_party\abseil-cpp -IT:\altsrc\github\grpc\workspace_c++_windows_dbg_native_default_default_buildonly\cmake\build\third_party\cares\cares -IT:\altsrc\github\grpc\workspace_c++_windows_dbg_native_default_default_buildonly\third_party\cares\cares -IT:\altsrc\github\grpc\workspace_c++_windows_dbg_native_default_default_buildonly\third_party\cares\cares\include /DWIN32 /D_WINDOWS /W3 /GR /EHsc  /MP8 /wd4065 /wd4506 /wd4200 /wd4291 /wd4244 /wd4267 /wd4987 /wd4774 /wd4819 /wd4996 /wd4619 /wd4503 /utf-8 /MDd /Z7 /Ob0 /Od /RTC1 /showIncludes /FoCMakeFiles\xds_listener_resource_type_test.dir\test\core\xds\xds_listener_resource_type_test.cc.obj /FdCMakeFiles\xds_listener_resource_type_test.dir\ /FS -c T:\altsrc\github\grpc\workspace_c++_windows_dbg_native_default_default_buildonly\test\core\xds\xds_listener_resource_type_test.cc
T:\altsrc\github\grpc\workspace_c++_windows_dbg_native_default_default_buildonly\third_party\protobuf\src\google/protobuf/arena_align.h(102): warning C4146: unary minus operator applied to unsigned type, result still unsigned
T:\altsrc\github\grpc\workspace_c++_windows_dbg_native_default_default_buildonly\third_party\protobuf\src\google/protobuf/arena_align.h(145): warning C4146: unary minus operator applied to unsigned type, result still unsigned
T:\altsrc\github\grpc\workspace_c++_windows_dbg_native_default_default_buildonly\src/core/lib/gprpp/time.h(314): warning C4146: unary minus operator applied to unsigned type, result still unsigned
T:\altsrc\github\grpc\workspace_c++_windows_dbg_native_default_default_buildonly\third_party\abseil-cpp\absl/types/compare.h(49): error C2338: Only literal `0` is allowed.
T:\altsrc\github\grpc\workspace_c++_windows_dbg_native_default_default_buildonly\third_party\abseil-cpp\absl/types/optional.h(704): note: see reference to class template instantiation 'absl::lts_20230125::compare_internal::Fail<T>' being compiled
        with
        [
            T=uint32_t

        ]

C:\Program Files (x86)\Microsoft Visual Studio\2017\Community\VC\Tools\MSVC\14.16.27023\include\xtr1common(87): note: see reference to class template instantiation 'std::integral_constant<bool,false>' being compiled

C:\Program Files (x86)\Microsoft Visual Studio\2017\Community\VC\Tools\MSVC\14.16.27023\include\xtr1common(88): note: see reference to class template instantiation 'std::is_same<_Ty1,_Ty2>' being compiled
```

https://source.cloud.google.com/results/invocations/cda777ff-cd4d-4c9e-a65f-77e6629313d6/targets/github%2Fgrpc%2Ftoplevel_run_tests_invocations%2Frun_tests_c%2B%2B_windows_dbg_native_default_default_buildonly/tests

--- a/absl/base/config.h
+++ b/absl/base/config.h
@@ -112,7 +112,7 @@
 // LTS releases can be obtained from
 // https://github.com/abseil/abseil-cpp/releases.
 #define ABSL_LTS_RELEASE_VERSION 20230125
-#define ABSL_LTS_RELEASE_PATCH_LEVEL 1
+#define ABSL_LTS_RELEASE_PATCH_LEVEL 2
 
 // Helper macro to convert a CPP variable to a string literal.
 #define ABSL_INTERNAL_DO_TOKEN_STR(x) #x

--- a/absl/types/compare.h
+++ b/absl/types/compare.h
@@ -44,29 +44,28 @@ namespace compare_internal {
 
 using value_type = int8_t;
 
-template <typename T>
-struct Fail {
-  static_assert(sizeof(T) < 0, "Only literal `0` is allowed.");
-};
+class OnlyLiteralZero {
+  // A private type which cannot be named to explicitly cast to it.
+  struct MatchLiteralZero;
 
-// We need the NullPtrT template to avoid triggering the modernize-use-nullptr
-// ClangTidy warning in user code.
-template <typename NullPtrT = std::nullptr_t>
-struct OnlyLiteralZero {
-  constexpr OnlyLiteralZero(NullPtrT) noexcept {}  // NOLINT
+ public:
+  // Accept only literal zero since it can be implicitly converted to a pointer
+  // type. nullptr constants will be caught by the other constructor which
+  // accepts a nullptr_t.
+  constexpr OnlyLiteralZero(MatchLiteralZero *) noexcept {}  // NOLINT
 
   // Fails compilation when `nullptr` or integral type arguments other than
   // `int` are passed. This constructor doesn't accept `int` because literal `0`
   // has type `int`. Literal `0` arguments will be implicitly converted to
   // `std::nullptr_t` and accepted by the above constructor, while other `int`
   // arguments will fail to be converted and cause compilation failure.
-  template <
-      typename T,
-      typename = typename std::enable_if<
-          std::is_same<T, std::nullptr_t>::value ||
-          (std::is_integral<T>::value && !std::is_same<T, int>::value)>::type,
-      typename = typename Fail<T>::type>
-  OnlyLiteralZero(T);  // NOLINT
+  template <typename T, typename = typename std::enable_if<
+                            std::is_same<T, std::nullptr_t>::value ||
+                            (std::is_integral<T>::value &&
+                             !std::is_same<T, int>::value)>::type>
+  OnlyLiteralZero(T) {  // NOLINT
+    static_assert(sizeof(T) < 0, "Only literal `0` is allowed.");
+  }
 };
 
 enum class eq : value_type {
@@ -163,18 +162,18 @@ class weak_equality
 
   // Comparisons
   friend constexpr bool operator==(
-      weak_equality v, compare_internal::OnlyLiteralZero<>) noexcept {
+      weak_equality v, compare_internal::OnlyLiteralZero) noexcept {
     return v.value_ == 0;
   }
   friend constexpr bool operator!=(
-      weak_equality v, compare_internal::OnlyLiteralZero<>) noexcept {
+      weak_equality v, compare_internal::OnlyLiteralZero) noexcept {
     return v.value_ != 0;
   }
-  friend constexpr bool operator==(compare_internal::OnlyLiteralZero<>,
+  friend constexpr bool operator==(compare_internal::OnlyLiteralZero,
                                    weak_equality v) noexcept {
     return 0 == v.value_;
   }
-  friend constexpr bool operator!=(compare_internal::OnlyLiteralZero<>,
+  friend constexpr bool operator!=(compare_internal::OnlyLiteralZero,
                                    weak_equality v) noexcept {
     return 0 != v.value_;
   }
@@ -214,18 +213,18 @@ class strong_equality
   }
   // Comparisons
   friend constexpr bool operator==(
-      strong_equality v, compare_internal::OnlyLiteralZero<>) noexcept {
+      strong_equality v, compare_internal::OnlyLiteralZero) noexcept {
     return v.value_ == 0;
   }
   friend constexpr bool operator!=(
-      strong_equality v, compare_internal::OnlyLiteralZero<>) noexcept {
+      strong_equality v, compare_internal::OnlyLiteralZero) noexcept {
     return v.value_ != 0;
   }
-  friend constexpr bool operator==(compare_internal::OnlyLiteralZero<>,
+  friend constexpr bool operator==(compare_internal::OnlyLiteralZero,
                                    strong_equality v) noexcept {
     return 0 == v.value_;
   }
-  friend constexpr bool operator!=(compare_internal::OnlyLiteralZero<>,
+  friend constexpr bool operator!=(compare_internal::OnlyLiteralZero,
                                    strong_equality v) noexcept {
     return 0 != v.value_;
   }
@@ -277,50 +276,50 @@ class partial_ordering
   }
   // Comparisons
   friend constexpr bool operator==(
-      partial_ordering v, compare_internal::OnlyLiteralZero<>) noexcept {
+      partial_ordering v, compare_internal::OnlyLiteralZero) noexcept {
     return v.is_ordered() && v.value_ == 0;
   }
   friend constexpr bool operator!=(
-      partial_ordering v, compare_internal::OnlyLiteralZero<>) noexcept {
+      partial_ordering v, compare_internal::OnlyLiteralZero) noexcept {
     return !v.is_ordered() || v.value_ != 0;
   }
   friend constexpr bool operator<(
-      partial_ordering v, compare_internal::OnlyLiteralZero<>) noexcept {
+      partial_ordering v, compare_internal::OnlyLiteralZero) noexcept {
     return v.is_ordered() && v.value_ < 0;
   }
   friend constexpr bool operator<=(
-      partial_ordering v, compare_internal::OnlyLiteralZero<>) noexcept {
+      partial_ordering v, compare_internal::OnlyLiteralZero) noexcept {
     return v.is_ordered() && v.value_ <= 0;
   }
   friend constexpr bool operator>(
-      partial_ordering v, compare_internal::OnlyLiteralZero<>) noexcept {
+      partial_ordering v, compare_internal::OnlyLiteralZero) noexcept {
     return v.is_ordered() && v.value_ > 0;
   }
   friend constexpr bool operator>=(
-      partial_ordering v, compare_internal::OnlyLiteralZero<>) noexcept {
+      partial_ordering v, compare_internal::OnlyLiteralZero) noexcept {
     return v.is_ordered() && v.value_ >= 0;
   }
-  friend constexpr bool operator==(compare_internal::OnlyLiteralZero<>,
+  friend constexpr bool operator==(compare_internal::OnlyLiteralZero,
                                    partial_ordering v) noexcept {
     return v.is_ordered() && 0 == v.value_;
   }
-  friend constexpr bool operator!=(compare_internal::OnlyLiteralZero<>,
+  friend constexpr bool operator!=(compare_internal::OnlyLiteralZero,
                                    partial_ordering v) noexcept {
     return !v.is_ordered() || 0 != v.value_;
   }
-  friend constexpr bool operator<(compare_internal::OnlyLiteralZero<>,
+  friend constexpr bool operator<(compare_internal::OnlyLiteralZero,
                                   partial_ordering v) noexcept {
     return v.is_ordered() && 0 < v.value_;
   }
-  friend constexpr bool operator<=(compare_internal::OnlyLiteralZero<>,
+  friend constexpr bool operator<=(compare_internal::OnlyLiteralZero,
                                    partial_ordering v) noexcept {
     return v.is_ordered() && 0 <= v.value_;
   }
-  friend constexpr bool operator>(compare_internal::OnlyLiteralZero<>,
+  friend constexpr bool operator>(compare_internal::OnlyLiteralZero,
                                   partial_ordering v) noexcept {
     return v.is_ordered() && 0 > v.value_;
   }
-  friend constexpr bool operator>=(compare_internal::OnlyLiteralZero<>,
+  friend constexpr bool operator>=(compare_internal::OnlyLiteralZero,
                                    partial_ordering v) noexcept {
     return v.is_ordered() && 0 >= v.value_;
   }
@@ -369,50 +368,50 @@ class weak_ordering
   }
   // Comparisons
   friend constexpr bool operator==(
-      weak_ordering v, compare_internal::OnlyLiteralZero<>) noexcept {
+      weak_ordering v, compare_internal::OnlyLiteralZero) noexcept {
     return v.value_ == 0;
   }
   friend constexpr bool operator!=(
-      weak_ordering v, compare_internal::OnlyLiteralZero<>) noexcept {
+      weak_ordering v, compare_internal::OnlyLiteralZero) noexcept {
     return v.value_ != 0;
   }
   friend constexpr bool operator<(
-      weak_ordering v, compare_internal::OnlyLiteralZero<>) noexcept {
+      weak_ordering v, compare_internal::OnlyLiteralZero) noexcept {
     return v.value_ < 0;
   }
   friend constexpr bool operator<=(
-      weak_ordering v, compare_internal::OnlyLiteralZero<>) noexcept {
+      weak_ordering v, compare_internal::OnlyLiteralZero) noexcept {
     return v.value_ <= 0;
   }
   friend constexpr bool operator>(
-      weak_ordering v, compare_internal::OnlyLiteralZero<>) noexcept {
+      weak_ordering v, compare_internal::OnlyLiteralZero) noexcept {
     return v.value_ > 0;
   }
   friend constexpr bool operator>=(
-      weak_ordering v, compare_internal::OnlyLiteralZero<>) noexcept {
+      weak_ordering v, compare_internal::OnlyLiteralZero) noexcept {
     return v.value_ >= 0;
   }
-  friend constexpr bool operator==(compare_internal::OnlyLiteralZero<>,
+  friend constexpr bool operator==(compare_internal::OnlyLiteralZero,
                                    weak_ordering v) noexcept {
     return 0 == v.value_;
   }
-  friend constexpr bool operator!=(compare_internal::OnlyLiteralZero<>,
+  friend constexpr bool operator!=(compare_internal::OnlyLiteralZero,
                                    weak_ordering v) noexcept {
     return 0 != v.value_;
   }
-  friend constexpr bool operator<(compare_internal::OnlyLiteralZero<>,
+  friend constexpr bool operator<(compare_internal::OnlyLiteralZero,
                                   weak_ordering v) noexcept {
     return 0 < v.value_;
   }
-  friend constexpr bool operator<=(compare_internal::OnlyLiteralZero<>,
+  friend constexpr bool operator<=(compare_internal::OnlyLiteralZero,
                                    weak_ordering v) noexcept {
     return 0 <= v.value_;
   }
-  friend constexpr bool operator>(compare_internal::OnlyLiteralZero<>,
+  friend constexpr bool operator>(compare_internal::OnlyLiteralZero,
                                   weak_ordering v) noexcept {
     return 0 > v.value_;
   }
-  friend constexpr bool operator>=(compare_internal::OnlyLiteralZero<>,
+  friend constexpr bool operator>=(compare_internal::OnlyLiteralZero,
                                    weak_ordering v) noexcept {
     return 0 >= v.value_;
   }
@@ -468,50 +467,50 @@ class strong_ordering
   }
   // Comparisons
   friend constexpr bool operator==(
-      strong_ordering v, compare_internal::OnlyLiteralZero<>) noexcept {
+      strong_ordering v, compare_internal::OnlyLiteralZero) noexcept {
     return v.value_ == 0;
   }
   friend constexpr bool operator!=(
-      strong_ordering v, compare_internal::OnlyLiteralZero<>) noexcept {
+      strong_ordering v, compare_internal::OnlyLiteralZero) noexcept {
     return v.value_ != 0;
   }
   friend constexpr bool operator<(
-      strong_ordering v, compare_internal::OnlyLiteralZero<>) noexcept {
+      strong_ordering v, compare_internal::OnlyLiteralZero) noexcept {
     return v.value_ < 0;
   }
   friend constexpr bool operator<=(
-      strong_ordering v, compare_internal::OnlyLiteralZero<>) noexcept {
+      strong_ordering v, compare_internal::OnlyLiteralZero) noexcept {
     return v.value_ <= 0;
   }
   friend constexpr bool operator>(
-      strong_ordering v, compare_internal::OnlyLiteralZero<>) noexcept {
+      strong_ordering v, compare_internal::OnlyLiteralZero) noexcept {
     return v.value_ > 0;
   }
   friend constexpr bool operator>=(
-      strong_ordering v, compare_internal::OnlyLiteralZero<>) noexcept {
+      strong_ordering v, compare_internal::OnlyLiteralZero) noexcept {
     return v.value_ >= 0;
   }
-  friend constexpr bool operator==(compare_internal::OnlyLiteralZero<>,
+  friend constexpr bool operator==(compare_internal::OnlyLiteralZero,
                                    strong_ordering v) noexcept {
     return 0 == v.value_;
   }
-  friend constexpr bool operator!=(compare_internal::OnlyLiteralZero<>,
+  friend constexpr bool operator!=(compare_internal::OnlyLiteralZero,
                                    strong_ordering v) noexcept {
     return 0 != v.value_;
   }
-  friend constexpr bool operator<(compare_internal::OnlyLiteralZero<>,
+  friend constexpr bool operator<(compare_internal::OnlyLiteralZero,
                                   strong_ordering v) noexcept {
     return 0 < v.value_;
   }
-  friend constexpr bool operator<=(compare_internal::OnlyLiteralZero<>,
+  friend constexpr bool operator<=(compare_internal::OnlyLiteralZero,
                                    strong_ordering v) noexcept {
     return 0 <= v.value_;
   }
-  friend constexpr bool operator>(compare_internal::OnlyLiteralZero<>,
+  friend constexpr bool operator>(compare_internal::OnlyLiteralZero,
                                   strong_ordering v) noexcept {
     return 0 > v.value_;
   }
-  friend constexpr bool operator>=(compare_internal::OnlyLiteralZero<>,
+  friend constexpr bool operator>=(compare_internal::OnlyLiteralZero,
                                    strong_ordering v) noexcept {
     return 0 >= v.value_;
   }

